### PR TITLE
[blog] Lexical and dynamic scope

### PR DESCRIPTION
**Pull request by mhyee**
This is the first of two blog posts that supersedes #202.

I decided to focus only on the definitions of lexical/dynamic scope, deferring all the R stuff to a later post. As a result, I significantly rewrote the definitions of lexical/dynamic scope, added some opinion about why I think it's so hard to get right, and then wrote a short timeline of lexical scope in ALGOL, Lisp, and Scheme.

For convenience, you can find a rendered version [here](http://www.ccs.neu.edu/home/mhyee/scratch/Lexical%20and%20Dynamic%20Scope.html?2).

I'll have a draft/PR for part two up soon. I'd like to get both posts published within days of each other, preferably before Wednesday next week.

### CODE REVIEW

File: blog/_src/posts/2019-08-14-lexical-and-dynamic-scope.md [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+also explored some history, we are ready to return to the original question.
+_Is R lexically or dynamically scoped?_
+
+In the next blog post, we'll answer that question, and also see how R can be
+very confusing.
+
+_I would like to thank Ben Greenman and Artem Pelenitsyn for their feedback on
+an earlier draft of this post._
+
+---
```
Comment (bennn) [inline blog/_src/posts/2019-08-14-lexical-and-dynamic-scope.md]: Delete this line and the heading below it?

File: blog/_src/posts/2019-08-14-lexical-and-dynamic-scope.md [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+if there was an "original" definition of lexical scope. I did not find one, but
+I was able to trace some of the connections between Lisp, Scheme, and ALGOL 60.
+This history is certainly incomplete, but I hope it is somewhat useful and
+interesting.
+
+  * **1960**. John McCarthy publishes the original paper on Lisp.[^2] In
+    _History of Lisp_,[^3] McCarthy writes that he borrowed the λ-notation from
+    Alonzo Church's lambda calculus, but none of the other ideas. He also
+    recounts an incident where a programmer desired lexical scoping, but Lisp
+    used dynamic scoping. This was later solved by the "FUNARG device."
```
Comment (bennn) [inline blog/_src/posts/2019-08-14-lexical-and-dynamic-scope.md]: Worth noting that on the same page, McCarthy says he considered dynamic scope a bug?

File: blog/_src/posts/2019-08-14-lexical-and-dynamic-scope.md (hunk 1 above)
Comment (mhyee) [reply blog/_src/posts/2019-08-14-lexical-and-dynamic-scope.md]: I'll keep the line and remove the heading. It looks weird in Markdown, but I think the line is necessary when rendered:

Line+heading:
image

Line only:
image

No line, no heading:
image

File: blog/_src/posts/2019-08-14-lexical-and-dynamic-scope.md (hunk 2 above)
Comment (mhyee) [reply blog/_src/posts/2019-08-14-lexical-and-dynamic-scope.md]: Excellent suggestion, thanks!

```diff
     recounts an incident where a programmer desired lexical scoping, but Lisp
-    used dynamic scoping. This was later solved by the "FUNARG device."
+    used dynamic scoping. McCarthy considered this to be a bug, which Steve
+    Russell later fixed by developing the "FUNARG device."
```

**Comment by mhyee** [comment]
Thanks! I don't want to merge yet, because I want to make sure the next post (which I still haven't started...) is done first.

File: blog/_src/posts/2019-08-14-lexical-and-dynamic-scope.md [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+explicitly destroyed. Furthermore, a later initialization hides an earlier one.
+
+
+### Commentary
+
+I found this approach very informative, because it explicitly distinguishes
+between space (scope) and time (extent), which further implies a separation
+between compile time and run time. This explains my unease with the definition
+of "dynamic scope"---it is nominally about textual regions in a program, but
+also requires consideration of run-time behaviour. Dynamic scope is a misnomer!
```
Comment (howell) [inline blog/_src/posts/2019-08-14-lexical-and-dynamic-scope.md]: damn english and its non-compositional compound words.

File: blog/_src/posts/2019-08-14-lexical-and-dynamic-scope.md [hunk 4]
```diff
@@ -0,0 +1,246 @@
...
+
+## What is scope?
+
+_Scope_ refers to the places in a program where a variable is visible and can
+be referenced.
+
+An interesting situation is when a function has free variables. Consider the
+example below:
+
+```r
```
Comment (howell) [inline blog/_src/posts/2019-08-14-lexical-and-dynamic-scope.md]: is this R syntax? Is it meant to be ambiguous whether `<-` could mean assignment?

File: blog/_src/posts/2019-08-14-lexical-and-dynamic-scope.md [hunk 5]
```diff
@@ -0,0 +1,246 @@
...
+Most programming languages we use today are lexically scoped. Intuitively, a
+human (or compiler) can determine the scope of a variable by just examining the
+source code of a program. In other words, a compiler can determine which
+_definition_ each variable refers to---but it may not be able to determine the
+_values_ of each variable.
+
+
+### Dynamic scoping
+
+Under _dynamic scoping_, a variable is bound to the most recent value assigned
```
Comment (howell) [inline blog/_src/posts/2019-08-14-lexical-and-dynamic-scope.md]: I find this talk of assignment a bit confusing. It's evocative of mutable state, which seems orthogonal to scoping.

**Comment by howell** [comment]
this got me wondering how I'd explain lexical & dynamic scope, and I wonder where your new insight would fault it:

Much like in algebra, programs use *names* to refer to the same thing. Programs *bind* a name to associate it with a value, then later *use* the name to refer to its value. Since it is possible to use a name erroneously (used but not bound, used before it is bound) or ambiguously (bound multiple times), we need some way of determining which binding a use refers to.

*Lexical* scoping describes strategies that resolve this question using the syntax of the program. Typically, it means a use refers to the nearest enclosing binding of the same name.

*Dynamic* scoping describes strategies that resolve this question using the runtime state of the program. Typically, it means name bindings create an entry in a runtime stack as they are evaluated, and a use refers to the topmost binding of the same name in the stack.

File: blog/_src/posts/2019-08-14-lexical-and-dynamic-scope.md (hunk 4 above)
Comment (mhyee) [reply blog/_src/posts/2019-08-14-lexical-and-dynamic-scope.md]: Yeah, `<-` is assignment in R. Originally I had more R examples.

I'll clarify the description below to actually say assignment.

(Or I could use `x = 1` because that's also valid R. ...along with `1 -> x`.)

File: blog/_src/posts/2019-08-14-lexical-and-dynamic-scope.md (hunk 5 above)
Comment (mhyee) [reply blog/_src/posts/2019-08-14-lexical-and-dynamic-scope.md]: This is tough. In the previous draft, I explained scoping from a "bottom up" perspective, i.e. variable lookup. But that bothered me, so I switched to a "top down," from the perspective of a variable definition (or assignment). It's still a bit weird, but I think it better fits the "scope is the places in a program where a variable reference is valid" story.

Later on, I try to explain why this definition of dynamic scope seems weird. I can try to clarify a little bit, a little earlier.

**Comment by mhyee** [comment]
> this got me wondering how I'd explain lexical & dynamic scope, and I wonder where your new insight would fault it:
> 
> Much like in algebra, programs use _names_ to refer to the same thing. Programs _bind_ a name to associate it with a value, then later _use_ the name to refer to its value. Since it is possible to use a name erroneously (used but not bound, used before it is bound) or ambiguously (bound multiple times), we need some way of determining which binding a use refers to.
> 
> _Lexical_ scoping describes strategies that resolve this question using the syntax of the program. Typically, it means a use refers to the nearest enclosing binding of the same name.
> 
> _Dynamic_ scoping describes strategies that resolve this question using the runtime state of the program. Typically, it means name bindings create an entry in a runtime stack as they are evaluated, and a use refers to the topmost binding of the same name in the stack.

Comparing this set of definitions to my set of definitions, there's the "bottom up, from variable lookup" and the "top down, from variable definition" perspectives. Also, I would say that your definition of lexical scope is stronger than mine, while your definition of dynamic scope is weaker than mine.

I think that in 95% of cases, your definitions and my definitions will agree. But they will disagree on R: by my definitions, R is lexically scoped, while by your definitions, R is dynamically scoped. (Or it's a bit of both.)

Variable lookup in R is (roughly) based on syntax, and the use of a name does refer to the nearest enclosing binding of the same name (unless you do funny business). But variable lookup also depends on the run-time state of the program, though it does not scan the call stack (unless you do funny business).

I would ask, how strict (or weak) should the definitions of lexical and dynamic scope be? What kinds of definitions are more useful?

I actually plan on discussing this in the second blog post, which I'm currently working on.

File: blog/_src/posts/2019-08-14-lexical-and-dynamic-scope.md (hunk 4 above)
Comment (howell) [reply blog/_src/posts/2019-08-14-lexical-and-dynamic-scope.md]: I'm not sure I was clear. In every language I know, the binding of `x` from `x <- 1` is in scope for the `x <- 2` line. As someone who has never written R, but thinks `<-` looks vaguely like mutation, `x <- 2` could either mean (1) imperatively update the value of `x` from `1` to `2`, or (2) introduce a new binding of `x`, potentially shadowing the earlier one.

If the answer is (1), the example is unambiguous and doesn't really say anything about lexical vs. dynamic scope. It should just yield `2`.

p.s. it seems like we understand "assignment" to mean different things. I think of it as imperative update of mutable state, whereas you seem to also use it to mean the association of a name (immutably) with a value, which I would call "binding".

**Comment by howell** [comment]
I'm not sure what you mean by strong/strict/weak in the terms of these definitions. I thought the definitions I gave include your definitions as specific strategies, but 

> they will disagree on R

seems to suggest otherwise, and I certainly don't know enough about R to say myself.

> variable lookup also depends on the run-time state of the program

If (the implementation of) variable lookup uses runtime state, but the result of it is predictable from the program text, I would call that lexical.

I think all of our definitions leave the door perfectly open to mixing lexical and dynamic scoping in the same program(ming language), as long as you can tell which variables are using which scoping. For a *single* variable, I think it makes less sense to have a hybrid approach, though the meaning of statements like "variables are lexically scoped up to the use of certain reflection and metaprogramming operations" should be clear (which is what I'm guessing is happening in R).

File: blog/_src/posts/2019-08-14-lexical-and-dynamic-scope.md (hunk 4 above)
Comment (mhyee) [reply blog/_src/posts/2019-08-14-lexical-and-dynamic-scope.md]: Thanks for the clarification. In this specific example, the answer is (2): `x <- 2` does not update the previous value of `x`; it introduces a new binding for `x` that shadows the previous one.

But the ambiguity here is that `<-` is called the assignment operator in R, and it can also do updates:

```r
x <- 1   # creates a new variable
x        # value is 1
f <- function() {
  x <- 2 # creates a new variable
  x      # value is 2
  x <- 3 # updates the variable created 2 lines above
  x      # value is 3
}
x        # value is 1, refers to the variable created on the first line
```

`<-` works by searching the current function's scope for a variable to update. If a variable is found, it updates it. Otherwise, it creates a new binding* in the current scope. `<-` can only mutate the current scope.

I think I'll rewrite this example to not use R, and I'll include a discussion on `<-` in the next blog post.

*Maybe I need a different word, because these "bindings" are mutable. Mapping?

**Comment by mhyee** [comment]
> I'm not sure what you mean by strong/strict/weak in the terms of these definitions. I thought the definitions I gave include your definitions as specific strategies

My intent was to define "lexical scoping" as something pretty general: the locations where a variable can be referenced is determined by the syntax of the program. My definition says nothing about name resolution.

On the other hand, my definition of "dynamic scoping" is very specific: it's about mutable state and the most recently assigned value. I interpret your definition as allowing other kinds of strategies.

> If (the implementation of) variable lookup uses runtime state, but the result of it is predictable from the program text, I would call that lexical.

Here's a really annoying example:

```r
x <- 1 # creates a mapping (x,1) at the top-level scope
f <- function(b) {
  if (b)
    x <- 2 # creates a mapping (x,2) within the current function's scope
  x # looks up x in the current function scope, if it doesn't find x, looks up x in the top-level scope
}
```

By looking at just the program text, I know that the third `x` refers to either the first or second `x`. But I won't know which one unless I know what argument `f` is called with.

> For a single variable, I think it makes less sense to have a hybrid approach, though the meaning of statements like "variables are lexically scoped up to the use of certain reflection and metaprogramming operations" should be clear (which is what I'm guessing is happening in R).

I agree, though for R, I would have to say something like "variables are lexically scoped up to the use of certain features." But I've talked to some people who don't like the "up to the use of X" and think that disqualifies R from being lexically scoped.

Or, there was a paper that said "scoping in R is dynamic." I'm not sure if they mean "the scoping and lookup rules are dynamic," which is a fair assessment, or if they mean "R is dynamically scoped, i.e. variables take on the value of the most recent assignment."

**Comment by howell** [comment]
> Here's a really annoying example:

Clearly dynamic IMO :-)

(when I say predictable, I mean perfectly predictable, as in you can determine exactly which binding a name refers to)

> But I've talked to some people who don't like the "up to the use of X" and think that disqualifies R from being lexically scoped.

As in, they want to say that R is lexically scoped? From this discussion that appears to be factually incorrect (well, since these definitions are somewhat up in the air, incorrect based on *my* facts). Is this some kind of R propaganda effort?

> Or, there was a paper that said "scoping in R is dynamic." I'm not sure if they mean "the scoping and lookup rules are dynamic," which is a fair assessment, or if they mean "R is dynamically scoped, i.e. variables take on the value of the most recent assignment."

Using my definition, it doesn't matter exactly which they mean, so I would give such a generic statement a seal of approval.

**Comment by mhyee** [comment]
> Clearly dynamic IMO :-)
> 
> (when I say predictable, I mean perfectly predictable, as in you can determine exactly which binding a name refers to)

And that's where we disagree.

> As in, they want to say that R is lexically scoped? From this discussion that appears to be factually incorrect (well, since these definitions are somewhat up in the air, incorrect based on my facts). Is this some kind of R propaganda effort?

I would say something like "R is lexically scoped, up to the use of X," and then they would object, saying "if there's any exceptions, then it isn't lexical scoping."

But I think it's still useful to have a statement like "R is lexically scoped, except for these cases ..."

My goal for this blog post was to present a set of definitions on scoping, which I hope we can all agree on. (Maybe we don't. Oh well.) My goal for the second post is to discuss R, a language that really messes with scoping definitions, and argue why it _is_ lexically scoped (at least, under the definitions I presented), why it might be useful to use those definitions (rather than ones that would make R dynamically scoped), and that if you're going to say that R is lexically (or dynamically) scoped, you should be precise about what you actually mean.

>> Or, there was a paper that said "scoping in R is dynamic." I'm not sure if they mean "the scoping and lookup rules are dynamic," which is a fair assessment, or if they mean "R is dynamically scoped, i.e. variables take on the value of the most recent assignment."
>
> Using my definition, it doesn't matter exactly which they mean, so I would give such a generic statement a seal of approval.

But the second meaning (specifically, "variables take on the value of the most recent assignment") is incorrect. Still, I think it's more helpful to be precise, especially if different people have different definitions for lexical vs dynamic scope.

**Comment by howell** [comment]
to be clear, I'm not advocating for any changes to the post besides clarifying where mutable updates are occurring, or potentially occurring.

And yeah it's not surprising to me at all that some people in our community cannot, or will not, grasp the difference between scientific classification (R is dynamically scoped in the technical sense) and human understanding (R acts like its lexically scoped up to a few caveats).

**Comment by mhyee** [comment]
Thanks, I'll add the clarification regarding mutable updates. I still think this was a productive discussion, and it'll help with the next blog post.

**Comment by mfelleisen** [comment]
I have scanned the comments on the blog post, not the post itself. I will say that it reminds me of the 1970s and 1980s when people were confused about what was what. Until I saw this discussion I firmly believed that we had resolved these items completely. Until I saw this discussion, I also thought that Brian Harvey was the last man in CS alive who defended the value of dynamic scoping (in the early 90s and got a rebuttal from me and SK); he apologized for being dense about it to me 25 years later in Amsterdam, probably because he's retired now. 

I'd definitely read up on the old material (starting with Guy's book on CL). 

I'd also spent some time on denotational semantics, which clarified many terms very precisely because -- unlike modern operational semantics (mea culpa, I know most of you use a flavor of what I created) -- it can cleanly separate phases and concepts (which is why it is harder to use and prove theorems with unless you like laborious stuff).

File: blog/_src/posts/2019-08-14-lexical-and-dynamic-scope.md (hunk 4 above)
Comment (mhyee) [reply blog/_src/posts/2019-08-14-lexical-and-dynamic-scope.md]: I ended up keeping the R syntax, but clarifying what was going on in the prose.

```diff
-In this snippet, we define a variable `x` at the top level and two functions
-`f` and `g`. The body of `f` uses the parameter `a`, but also the free variable
-`x`. On line 7, we call `g`, which in turn calls `f` and returns that result.
+On line 1, we create a mapping for `x` with value `1`. On line 2, we define a
+function `f` whose body uses the parameter `a`, but also the free variable `x`.
+On line 3, we define a function `g`, whose body creates a new mapping for `x`
+with value `2`, and then calls `f(0)`. (Note that line 4 does not update the
+mapping created on line 1.) Finally, on line 7, we call `g()`.
 
-What value does `g` return when it is called? On line 2, what value should the
-free variable `x` be bound to?
+What value does `g` return when it is called? What mapping does the free
+variable `x` on line 2 refer to? Does it refer to the mapping on line 1 that
+was visible when `f` was defined? Or does it refer to the mapping on line 4
+that was created just before `f` was called?
```

**Comment by mhyee** [comment]
Thanks again everyone for the feedback. I will publish this tomorrow morning, and then parts two and three next week.

Final rendered snapshot [here](http://www.ccs.neu.edu/home/mhyee/scratch/Lexical%20and%20Dynamic%20Scope.html?2).
